kubernetes/examples/FasterTransformer_HuggingFace_Bert.md

```diff
@@ -0,0 +1,279 @@
+```
+
+## Copy Mar file and config.properties to PVC
+
+```bash
+kubectl exec --tty pod/model-store-pod -- mkdir /pv/model-store/
+kubectl cp BERTSeqClassification.mar model-store-pod:/pv/model-store/BERTSeqClassification.mar
+
+kubectl exec --tty pod/model-store-pod -- mkdir /pv/config/
+kubectl cp config.properties model-store-pod:/pv/config/config.properties
+```
+
+## Build Torchserve image
+
+1. Clone Torchserve Repo
+
+```bash
+git clone https://github.com/pytorch/serve.git
+cd serve/docker
+```
+
+2. Modify Python and Pip paths in `Dockerfile` as below
+
+```bash
+sed -i 's#/usr/bin/python3#/opt/conda/bin/python3#g' Dockerfile
+sed -i 's#/usr/local/bin/pip3#/opt/conda/bin/pip3#g' Dockerfile
+```
+
+3. Change GPU check in `Dockerfile` for nvcr.io image
+
+```bash
+sed -i 's#grep -q "cuda:"#grep -q "nvidia:"#g' Dockerfile
+```
+
+4. Build image
+
+```bash
+DOCKER_BUILDKIT=1 docker build -file Dockerfile -t <image-name> --build-arg BASE_IMAGE=nvcr.io/nvidia/pytorch:20.12-py3 --build-arg CUDA_VERSION=cu102 .
+```
+
+5. Push image
+
+```bash
+docker push <image-name>
+```
+
+## Install Torchserve
+
+1. Navigate to kubernetes TS Helm package folder
+
+```bash
+cd ../kubernetes/Helm
+```
+
+2. Modify values.yaml with image and memory
+
+```yaml
+torchserve_image: <image build in previous step>
+
+namespace: torchserve
+
+torchserve:
+  management_port: 8081
+  inference_port: 8080
+  metrics_port: 8082
+  pvd_mount: /home/model-server/shared/
+  n_gpu: 1
+  n_cpu: 4
+  memory_limit: 32Gi
+  memory_request: 32Gi
+
+deployment:
+  replicas: 1
+
+persitant_volume:
+  size: 1Gi
+```
+
+3. Install TS
+
+```bash
+helm install torchserve .
+```
+
+4. Check TS installation
+
+```bash
+Kubectl get pods -n default
+Kubectl logs <pod-name> -n default
+```
+
+## Run Inference
+
+1. Start a shell session into the TS pod
+
+```bash
+kubectl exec -it <pod-name> -- bash
+```
+
+2. Create a file curl-format.txt with the content below
+
+```bash
+    time_namelookup:    %{time_namelookup}s\n
+    time_connect:       %{time_connect}s\n
+    time_appconnect:    %{time_appconnect}s\n
+    time_pretransfer:   %{time_pretransfer}s\n
+    time_redirect:      %{time_redirect}s\n
+
+    time_starttransfer: %{time_starttransfer}s\n
+                        ----------\n
+    time_total:         %{time_total}s\n
+```
+
+3. Create input file
+
+Sample_text_captum_input.txt
+
+```json
+{
+  "text": "Bloomberg has decided to publish a new report on the global economy.",
+  "target": 1
+}
+```
+
+4. Run inference
+
+curl -w "@curl-format.txt" -o /dev/null -s -X POST http://127.0.0.1:8080/predictions/bert -T Seq_classification_artifacts/sample_text_captum_input.txt
+
+Output:
+
+```bash
+time_namelookup:    0.000021s
+time_connect:       0.000078s
+time_appconnect:    0.000000s
+time_pretransfer:   0.000096s
+time_redirect:      0.000000s
+time_starttransfer: 0.000922s
+                    ----------
+time_total: 0.007053s
```

Comment:
Does this measurement include preprocessing, inference and post processing from the handler? I don't have much experience with K8 but would be curious more to learn about how stuff is benchmarked here

Reply:
This just give a high level info.